### PR TITLE
[MIR] Initial implementation for translating calls.

### PR DESCRIPTION
**Pull request by luqmana**
#29575

### CODE REVIEW

File: src/librustc_trans/trans/mir/constant.rs [hunk 1]
```diff
@@ -49,8 +68,11 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
     {
         let constant_ty = bcx.monomorphize(&constant.ty);
         match constant.literal {
-            mir::Literal::Item { .. } => {
-                unimplemented!()
+            mir::Literal::Item { def_id, substs } => {
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/constant.rs]: Wait until https://github.com/rust-lang/rust/pull/29907 is landed?

Comment (luqmana) [inline src/librustc_trans/trans/mir/constant.rs]: Yea sure, ended up seeing it later and it's definitely more complete. I'll rebase on top of that instead.

File: src/test/run-pass/mir_trans_calls.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        (a, b, c)
+    }
+    callee(a, b, c)
+}
+
+#[rustc_mir]
+fn test2(a: isize) -> isize {
+    // Test passing a single argument.
+    // Not using out pointer.
+    fn callee(a: isize) -> isize {
```
Comment (nikomatsakis) [inline src/test/run-pass/mir_trans_calls.rs]: can you add some tests using trait dispatch? (or does that not work yet)

Comment (luqmana) [inline src/test/run-pass/mir_trans_calls.rs]: Added some more tests for calling methods in different ways. #29907 + this pull also allow calling trait static methods which i've also included a test for.

**Comment by nikomatsakis** [comment]
So #29907 has landed. I see that this PR was rebased, presumably to account for that?

**Comment by nikomatsakis** [comment]
@bors r+

Thanks @luqmana! Exciting. This is a big hurdle.

**Comment by dotdash** [comment]
@luqmana That failure on windows is legit. The call to `callee` in `test1` moves the Box, but still drops it. Valgrind also complains about that on linux.

```
==9625==    at 0x148878: je_valgrind_freelike_block (valgrind.c:33)
==9625==    by 0x10CE9F: heap::deallocate::ha08705e845151bb14ea (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CE44: heap::exchange_free::h2c2ccc188aeabd2akga (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CE11: Box$LT$i32$GT$::drop.3501::h2325578bde125039 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CD31: test1::h2cd52b4aff680550eaa (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10D219: main::h31e20355ce7fed41xda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x115704: sys_common::unwind::try::try_fn::h7505552440393768204 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1132E8: __rust_try (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1153A6: rt::lang_start::h347da8ae9f69d42864x (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10F659: main (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==  Address 0x6021048 is 0 bytes inside a block of size 8 free'd
==9625==    at 0x148878: je_valgrind_freelike_block (valgrind.c:33)
==9625==    by 0x10CE9F: heap::deallocate::ha08705e845151bb14ea (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CE44: heap::exchange_free::h2c2ccc188aeabd2akga (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CE11: Box$LT$i32$GT$::drop.3501::h2325578bde125039 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CD27: test1::h2cd52b4aff680550eaa (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10D219: main::h31e20355ce7fed41xda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x115704: sys_common::unwind::try::try_fn::h7505552440393768204 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1132E8: __rust_try (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1153A6: rt::lang_start::h347da8ae9f69d42864x (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10F659: main (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==  Block was alloc'd at
==9625==    at 0x11FBD3: je_mallocx (jemalloc.c:1560)
==9625==    by 0x10E4CF: heap::allocate::habdc17f0007099bfPda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10E442: heap::exchange_malloc::h8c52e41374611846Jfa (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10D1DB: main::h31e20355ce7fed41xda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x115704: sys_common::unwind::try::try_fn::h7505552440393768204 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1132E8: __rust_try (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1153A6: rt::lang_start::h347da8ae9f69d42864x (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10F659: main (in /home/doener/src/rust-origin/mir_trans_calls)
==9625== 
==9625== Invalid read of size 4
==9625==    at 0x10E64F: cmp::impls::i32.PartialEq::eq::h6dd47283b5ba64938fA (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10E632: boxed::Box$LT$T$GT$.PartialEq::eq::eq::h15679945919347397840 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10E589: tuple::_$LP$A$C$$u20$B$C$$u20$C$RP$.PartialEq::eq::eq::h13937896417118323772 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10D3FD: main::h31e20355ce7fed41xda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x115704: sys_common::unwind::try::try_fn::h7505552440393768204 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1132E8: __rust_try (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1153A6: rt::lang_start::h347da8ae9f69d42864x (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10F659: main (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==  Address 0x6021048 is 0 bytes inside a block of size 8 free'd
==9625==    at 0x148878: je_valgrind_freelike_block (valgrind.c:33)
==9625==    by 0x10CE9F: heap::deallocate::ha08705e845151bb14ea (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CE44: heap::exchange_free::h2c2ccc188aeabd2akga (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CE11: Box$LT$i32$GT$::drop.3501::h2325578bde125039 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10CD27: test1::h2cd52b4aff680550eaa (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10D219: main::h31e20355ce7fed41xda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x115704: sys_common::unwind::try::try_fn::h7505552440393768204 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1132E8: __rust_try (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1153A6: rt::lang_start::h347da8ae9f69d42864x (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10F659: main (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==  Block was alloc'd at
==9625==    at 0x11FBD3: je_mallocx (jemalloc.c:1560)
==9625==    by 0x10E4CF: heap::allocate::habdc17f0007099bfPda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10E442: heap::exchange_malloc::h8c52e41374611846Jfa (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10D1DB: main::h31e20355ce7fed41xda (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x115704: sys_common::unwind::try::try_fn::h7505552440393768204 (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1132E8: __rust_try (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x1153A6: rt::lang_start::h347da8ae9f69d42864x (in /home/doener/src/rust-origin/mir_trans_calls)
==9625==    by 0x10F659: main (in /home/doener/src/rust-origin/mir_trans_calls)
```

**Comment by nagisa** [comment]
See https://github.com/rust-lang/rust/issues/29855

**Comment by dotdash** [comment]
As @nagisa said on IRC, I think it's fine to just remove the `Box` argument for now to make the test work.

**Comment by nikomatsakis** [comment]
On Fri, Dec 18, 2015 at 08:20:02AM -0800, Simonas Kazlauskas wrote:
> See https://github.com/rust-lang/rust/issues/29855

Oh yeah, we probably want to modify trans to zero. Bah. Forgot about
that!

**Comment by luqmana** [comment]
@bors: r=nikomatsakis

Just removed Box from the test for now til we get drop working.

**Pull request merged**